BE-6-add-auth0-and-jwt-auth

Randayy opened the pull request:
<img width="1015" alt="image" src="https://github.com/Randayy/FastAPI_App/assets/108235770/d6dee84c-8157-418a-9dd3-767d27fba98f">
<img width="1015" alt="image" src="https://github.com/Randayy/FastAPI_App/assets/108235770/ecb7dda5-dfaf-4a1c-806a-a9c35b5118fa">
<img width="1015" alt="image" src="https://github.com/Randayy/FastAPI_App/assets/108235770/92accd29-a62b-416a-8957-b9cf6ede2526">
<img width="1678" alt="image" src="https://github.com/Randayy/FastAPI_App/assets/108235770/504ed834-4f63-426e-9f89-949641d6da77">

Randayy committed:
added jwt auth getting token

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+from typing import Annotated
+
+
+SECRET_KEY = 'FE3FDF34FE3FF34F34F34F34F'
```
move SECRET_KEY to env and then take it from the settings

Randayy replied:
moved

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+
+
+SECRET_KEY = 'FE3FDF34FE3FF34F34F34F34F'
+ALGORITHM = 'HS256'
```
add new variable in the settings JWT_AUTH_ ALGORITHM

Randayy replied:
added

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+
+SECRET_KEY = 'FE3FDF34FE3FF34F34F34F34F'
+ALGORITHM = 'HS256'
+ACCESS_TOKEN_EXPIRE_MINUTES = 30
```
move to the settings

Randayy replied:
moved

Artemooon commented on app/db/auth_models.py:
```diff
@@ -0,0 +1,11 @@
+from sqlalchemy import Column, String
```
auth_schemas

Randayy replied:
renamed

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+        current_password = user.password
+        if not verify_password(password, current_password):
+            raise HTTPException(status_code=400, detail="Incorrect password")
+        return user
```
move this method to the repository level user_repository or auth_repository

Randayy replied:
I moved it to user service and called get_user_by_username from repository

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+
+
+class JWTAuth():
+    def __init__(self, db: AsyncSession):
```
remove db from here

Randayy replied:
removed

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+        to_encode = data.copy()
+        if self.access_token_expire_minutes:
+            expire = datetime.utcnow() + timedelta(minutes=self.access_token_expire_minutes)
+        else:
```
why do we need else here? is it that access_token_expire_minutes will no exist?

Randayy replied:
deleted else

Randayy committed:
added /me

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+            raise HTTPException(status_code=400, detail="Incorrect password")
+        return user
+
+    async def create_access_token(self, data: dict):
```
use pydantic sheme here to specify data that we expect here

Randayy replied:
used

Artemooon commented on app/routers/user_routes.py:
```diff
@@ -45,3 +46,11 @@ async def update_user(user_id: UUID, user_data: UserUpdateRequestSchema, db: Asy
+    auth = JWTAuth(db)
+    user = await auth.authenticate_user(form_data.username, form_data.password)
+    access_token = await auth.create_access_token(data={"sub": user.username})
+    return {"access_token": access_token, "token_type": "bearer"}
```
return Token(access_token= access_token, token_type="bearer")

Randayy replied:
made

Artemooon commented on app/routers/user_routes.py:
```diff
@@ -45,3 +46,17 @@ async def update_user(user_id: UUID, user_data: UserUpdateRequestSchema, db: Asy
+    return {"access_token": access_token, "token_type": "bearer"}
+
+@user_router.get("/me", response_model=UserDetailSchema)
+async def read_users_me(token: str, db: AsyncSession = Depends(get_session)):
```
here you should take your token from the Authorization header. the frontend will send this format 
Authorization: Bearer <token>

Randayy replied:
fixed

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+            username: str = payload.get("sub")
+            if username is None:
+                return None
+        except jwt.JWTError:
```
do you handle the case when token already expired?

Randayy replied:
yes

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+                return None
+        except jwt.JWTError:
+            raise HTTPException(
+                status_code=403, detail="Could not validate credentials")
```
I would return 401 here

Randayy replied:
changed

Artemooon commented on app/auth/jwtauth.py:
```diff
@@ -0,0 +1,59 @@
+                                 algorithms=[self.algorithm])
+            username: str = payload.get("sub")
+            if username is None:
+                return None
```
here also return 401 code error

Randayy replied:
changed

Artemooon commented on app/db/auth_models.py:
```diff
@@ -0,0 +1,11 @@
+
+
+class TokenData(BaseModel):
+    username: str | None = None
```
add exp for expire time and also sub - user id and username - username

Randayy replied:
added

Randayy replied:
user id I can't do because I got error while adding that Object of type UUID is not JSON serializable

Randayy committed:
fix

Randayy committed:
try

Randayy committed:
fixes

Randayy committed:
Update jwtauth.py

Artemooon commented on app/schemas/auth_schemas.py:
```diff
@@ -0,0 +1,14 @@
+class TokenData(BaseModel):
+    sub: str | None = None
+    exp: int | None = None
+    # sub: UUID | None = None
```
commented code

Randayy replied:
deleted

Artemooon commented on app/services/user_service.py:
```diff
@@ -46,23 +60,56 @@ async def update_user(self, user_id: UUID, user_data: UserUpdateRequestSchema) -
+            raise HTTPException(
+                status_code=400, detail="user with email already exists")
 
+        print(current_password, entered_password)
```
remove print

Randayy replied:
removed

Randayy committed:
fix auth jwt finally

Randayy committed:
try

Randayy committed:
formated

Randayy commented:
<img width="1501" alt="image" src="https://github.com/Randayy/FastAPI_App/assets/108235770/c719295f-da2f-4ad0-8600-201df864fcf0">

Artemooon commented on app/services/user_service.py:
```diff
@@ -46,23 +62,98 @@ async def update_user(self, user_id: UUID, user_data: UserUpdateRequestSchema) -
+            new_user_data = {
+                "username": current_user.get("email").split("@")[0],
+                "email": current_user.get("email"),
+                "password": "testpassword",
```
generate strong random password here

Artemooon commented on app/services/user_service.py:
```diff
@@ -46,23 +62,98 @@ async def update_user(self, user_id: UUID, user_data: UserUpdateRequestSchema) -
+            raise HTTPException(
+                status_code=401, detail="Token has expired")
+        user = await self.user_repository.get_user_by_username(username=token_data.sub)
+        if user is None:
```
if not user

Artemooon commented on app/auth/auth0.py:
```diff
@@ -0,0 +1,41 @@
+#     data = res.read()
+
+
+#     return data.decode("utf-8")
```
commented code? remove it

Randayy replied:
deleted this file is not needed

Artemooon commented on app/routers/user_routes.py:
```diff
@@ -45,3 +47,28 @@ async def update_user(user_id: UUID, user_data: UserUpdateRequestSchema, db: Asy
+    return current_user
+
+
+@user_router.get("/auth0/me")
```
use single router for getting current user, pass the token type to the payload and check if it's token created by you or by auth0

Randayy replied:
created

Artemooon commented on app/services/user_service.py:
```diff
@@ -46,23 +62,98 @@ async def update_user(self, user_id: UUID, user_data: UserUpdateRequestSchema) -
+        current_user = await self.get_current_user(token)
+        return current_user
+
+    async def get_current_user_from_token(self, token: HTTPAuthorizationCredentials) -> UserDetailSchema:
```
move this logic to a universal method for getting the user

Randayy committed:
fixed into 1 /me

i created 2 endpoints for 2 different authorizations , it was my bad i did not correctly understand task

Randayy committed:
password gen
